contracts/src/proxy/utils/uups_upgradeable.rs

```diff
@@ -0,0 +1,743 @@
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for UUPSUpgradeable {}
+
+/// Interface for a UUPS (Universal Upgradeable Proxy Standard) upgradeable
+/// contract.
+#[interface_id]
+pub trait IUUPSUpgradeable: IErc1822Proxiable {
+    /// Returns the version of the upgrade interface of the contract.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    fn upgrade_interface_version(&self) -> String {
+        UPGRADE_INTERFACE_VERSION.into()
+    }
+
+    /// Upgrade the implementation of the proxy to `new_implementation`, and
+    /// subsequently execute the function call encoded in `data`.
+    ///
+    /// Note: This function should revert when [`stylus_sdk::msg::sender`] is
+    /// not authorized to upgrade the contract.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `new_implementation` - The address of the new implementation contract.
+    /// * `data` - Additional data to be passed to the new implementation.
+    ///
+    /// # Errors
+    ///
+    /// * [`Error::UnauthorizedCallContext`] - If the call is not made through a
+    ///   valid proxy context.
+    /// * [`Error::InvalidImplementation`] - If the new implementation address
+    ///   is invalid or doesn't implement the required interface.
+    /// * [`Error::UnsupportedProxiableUUID`] - If the new implementation
+    ///   returns an unsupported UUID.
+    /// * [`Error::NonPayable`] - If the upgrade function receives ETH but is
+    ///   not designed to handle it.
+    /// * [`Error::EmptyCode`] - If there's no code at the new implementation
+    ///   address.
+    /// * [`Error::FailedCall`] - If the [`stylus_sdk::call::delegate_call`] to
+    ///   the new implementation fails.
+    /// * [`Error::FailedCallWithReason`] - If the
+    ///   [`stylus_sdk::call::delegate_call`] fails with a specific reason.
+    ///
+    /// # Events
+    ///
+    /// * [`crate::proxy::erc1967::Upgraded`]: Emitted when the implementation
+    ///   is upgraded.
+    fn upgrade_to_and_call(
+        &mut self,
+        new_implementation: Address,
+        data: Bytes,
+    ) -> Result<(), Vec<u8>>;
+}
+
+/// State of a [`UUPSUpgradeable`] contract.
+#[storage]
+pub struct UUPSUpgradeable {
+    /// The address of this contract, used for context validation.
+    pub self_address: StorageAddress,
+}
+
+#[public]
+#[implements(IUUPSUpgradeable, IErc1822Proxiable)]
+impl UUPSUpgradeable {
+    /// Initializes the contract by storing its own address for later context
+    /// validation.
+    ///
+    /// Ugly hack with setting the `self_address` storage value.
+    ///
+    /// Stylus SDK doesn't support setting the immutable storage values as
+    /// in Solidity:
+    ///
+    /// ```solidity
+    /// address private immutable __self = address(this);
+    /// ```
```

Comment:
```suggestion
    /// Unlike Solidity's immutable variables, Stylus requires storing the
    /// contract address in a storage field. This additional storage slot
    /// enables the same upgrade safety checks as the Solidity implementation
    /// without affecting the contract's upgrade behavior.
```
Stating that our implementation uses an "ugly hack" won't inspire confidence in devs. The updated description is just as clear and sounds more professional.